apollo-router/src/files.rs

```diff
@@ -1,48 +1,56 @@
 /// and must be dropped to finish watching.
 ///
 /// # Arguments
 ///
 /// * `path`: The file to watch
-/// * `delay`: The delay before sending an event. This prevents duplicate events for large writes. Defaults to 2 seconds.
 ///
 /// returns: impl Stream<Item=()>
 ///
-pub(crate) fn watch(path: PathBuf, delay: Option<Duration>) -> impl Stream<Item = ()> {
+pub(crate) fn watch(path: &Path) -> impl Stream<Item = ()> {
     let (mut watch_sender, watch_receiver) = mpsc::channel(1);
     let mut watcher =
-        Hotwatch::new_with_custom_delay(delay.unwrap_or_else(|| Duration::from_secs(2)))
-            .expect("Failed to initialise file watching.");
-    watcher
-        .watch(path, move |event: hotwatch::Event| {
-            tracing::debug!("file watcher: received event: {:?}", &event);
-            match event {
-                // https://github.com/notify-rs/notify/blob/ded07f442a96f33c6b7fefe3195396a33a28ddc3/src/lib.rs#L413
-                //
-                // `Create` events have a higher priority than `Write` and `Chmod`. These events will not be
-                // emitted if they are detected before the `Create` event has been emitted.
-                //
-                // Hotwatch will sometimes send a Create event instead of a Write event,
-                // if the write occured immediately after the file creation.
-                hotwatch::Event::Write(_) | hotwatch::Event::Create(_) => {
-                    if let Err(_err) = watch_sender.try_send(()) {
-                        tracing::error!(
-                            "Failed to process file watch notification. {}",
-                            _err.to_string()
-                        )
+        notify::recommended_watcher(move |res: Result<notify::Event, notify::Error>| match res {
+            Ok(event) => {
+                // We are only interested in  modify events.
+                // We don't want to lose events and a slow consumer could make us
+                // miss an event notification without a re-send strategy.
+                // If we can't send the event because the channel is full, wait
+                // for a short while and try again. Otherwise, we will panic
+                // because it's a non-recoverable error.
+                if let notify::event::EventKind::Modify(_) = event.kind {
+                    loop {
+                        match watch_sender.try_send(()) {
+                            Ok(_) => break,
+                            Err(err) => {
+                                tracing::warn!(
+                                    "could not process file watch notification. {}",
+                                    err.to_string()
+                                );
+                                if err.is_full() {
+                                    std::thread::sleep(Duration::from_millis(50));
+                                } else {
+                                    panic!("event channel failed: {}", err);
```

Comment:
uh why is there a panic here ? ^^'

Reply:
If we try to send a message and the channel is disconnected, that's a programming logic bug and so panic seems appropriate.

Reply:
I guess I'm not sure what conditions disconnected could occur in. I suppose if it happened during shutdown we would just want to silently ignore the fact that the event wasn't delivered, but can it happen at any other time?

Reply:
I see, the errors are either Full or Disconnected, and Disconnected means we have bigger issues because the state machine is botched somehow ?

Reply:
I don't actually think Disconnected can happen in our use case because we push the watcher into the stream and (as the comment below says) it will never be dropped. That means that I don't think this code will ever be triggered unless a programming logic error is introduced. In which case, panic is the correct response.

Reply:
I see, LGTM!